src/System.Globalization.Calendars/tests/CalendarHelpers.cs

```diff
@@ -2,177 +2,411 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
 using Xunit;
 
 namespace System.Globalization.Tests
 {
     public static class CalendarHelpers
     {
-        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
+        public enum DataType
+        {
+            Year = 1,
+            Month = 2,
+            Day = 8
+        }
+        
+        public static Calendar[] s_calendars = new Calendar[]
+        {
+            new ThaiBuddhistCalendar(),
+            new JapaneseCalendar(),
+            new KoreanCalendar(),
+            new TaiwanCalendar(),
+            new GregorianCalendar(GregorianCalendarTypes.Arabic),
+            new GregorianCalendar(GregorianCalendarTypes.Localized),
+            new GregorianCalendar(GregorianCalendarTypes.MiddleEastFrench),
+            new GregorianCalendar(GregorianCalendarTypes.TransliteratedEnglish),
+            new GregorianCalendar(GregorianCalendarTypes.TransliteratedFrench),
+            new GregorianCalendar(GregorianCalendarTypes.USEnglish),
+            new HijriCalendar(),
+            new HebrewCalendar(),
+            new JulianCalendar(),
+            new TaiwanLunisolarCalendar(),
+            new ChineseLunisolarCalendar(),
+            new KoreanLunisolarCalendar(),
+            new PersianCalendar()
+        };
+
         private static int MinEra(Calendar calendar) => calendar.GetEra(calendar.MinSupportedDateTime);
+        private static int MaxEra(Calendar calendar) => calendar.GetEra(calendar.MaxSupportedDateTime);
 
-        private static int YearOffset(Calendar calendar) => calendar.GetYear(calendar.MinSupportedDateTime) - 1;
+        private static int MaxCalendarYearInEra(Calendar calendar, int era)
+        {
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[0] || era == 0)
+            {
+                return calendar.GetYear(calendar.MaxSupportedDateTime);
+            }
+            return calendar.GetYear(calendar.ToDateTime(1, 1, 1, 0, 0, 0, 0, era + 1).AddDays(-1)) + 1;
+        }
 
-        public static void AddYears_Invalid(Calendar calendar)
+        private static int MaxGregorianYearInEra(Calendar calendar, int era)
         {
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MaxSupportedDateTime, 1));
-            Assert.Throws<ArgumentOutOfRangeException>(() => calendar.AddYears(calendar.MinSupportedDateTime, -1));
+            int[] eras = calendar.Eras;
+            Assert.InRange(era, 0, eras[0]);
+            if (eras.Length == 1 || era == eras[0] || era == 0)
+            {
+                return calendar.MaxSupportedDateTime.Year;
+            }
+            return MinGregorianYearInEra(calendar, era + 1);
```

Comment:
this one should be calendar.ToDateTime(1, 1, 1, 0, 0, 0, 0, era + 1).AddDays(-1)).Year.